File: Vialer/Availability/Controllers/AvailabilityViewController.m
//
//  AvailabilityViewController.m
//  Copyright (c) 2015 VoIPGRID. All rights reserved.
//

#import "AvailabilityViewController.h"

#import "AvailabilityModel.h"
#import "Configuration.h"
#import "SVProgressHUD.h"
#import "UIAlertController+Vialer.h"
#import "Vialer-Swift.h"
#import "VialerWebViewController.h"
#import "VoIPGRIDRequestOperationManager.h"

@interface AvailabilityViewController()
@property (nonatomic, weak) NSIndexPath *lastSelected;
@property (nonatomic, strong) AvailabilityModel *availabilityModel;
@end

static NSString * const AvailabilityAddFixedDestinationSegue = @"AddFixedDestinationSegue";
static NSString * const AvailabilityViewControllerAddFixedDestinationPageURLWithVariableForClient = @"/client/%@/fixeddestination/add/";

@implementation AvailabilityViewController

- (void)awakeFromNib {
    [super awakeFromNib];
     self.refreshControl.attributedTitle = [[NSAttributedString alloc] initWithString: NSLocalizedString(@"Loading availability options...", nil)];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [VialerGAITracker trackScreenForControllerWithName:NSStringFromClass([self class])];

    [self.refreshControl beginRefreshing];
    [self.tableView setContentOffset:CGPointMake(0, -self.refreshControl.frame.size.height) animated:YES];
    [self loadUserDestinationsWithRefreshControl:self.refreshControl];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.tableView.tintColor = [[Configuration defaultConfiguration].colorConfiguration colorForKey:ConfigurationAvailabilityTableViewTintColor];
}

- (AvailabilityModel *)availabilityModel {
    if (_availabilityModel == nil) {
        _availabilityModel = [[AvailabilityModel alloc] init];
    }
    return _availabilityModel;
}

- (IBAction)loadUserDestinationsWithRefreshControl:(UIRefreshControl *)sender {
    [self.availabilityModel getUserDestinations:^(NSString *localizedErrorString) {
        if (localizedErrorString) {
            [self presentViewController:[UIAlertController alertControllerWithTitle:NSLocalizedString(@"Error", nil)
                                                                            message:localizedErrorString
                                                               andDefaultButtonText:NSLocalizedString(@"Ok", nil)]
                               animated:YES
                             completion:nil];
        }else{
            [self.tableView reloadData];
        }
        [self.refreshControl endRefreshing];
    }];
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.availabilityModel.availabilityOptions count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *DefaultCellIdentifier = @"AvailabilityTableViewDefaultCell";
    NSDictionary *availabilityDict = self.availabilityModel.availabilityOptions[indexPath.row];

    UITableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:DefaultCellIdentifier];
    if ([availabilityDict[AvailabilityModelPhoneNumberKey] isEqualToNumber:@0]){
        cell.textLabel.text = availabilityDict[AvailabilityModelDescription];
    } else {
        NSString *phoneNumber = [availabilityDict[AvailabilityModelPhoneNumberKey] stringValue];
        if (phoneNumber.length > 5) {
            phoneNumber = [@"+" stringByAppendingString:phoneNumber];
        }
        cell.textLabel.text = [NSString stringWithFormat:@"%@ / %@",  phoneNumber, availabilityDict[AvailabilityModelDescription]];
    }

    if ([availabilityDict[AvailabilityModelSelected] isEqualToNumber:[NSNumber numberWithBool:YES]]) {
        cell.accessoryType = UITableViewCellAccessoryCheckmark;
        self.lastSelected = indexPath;
    } else {
        cell.accessoryType = UITableViewCellAccessoryNone;
    }

    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    [tableView cellForRowAtIndexPath:self.lastSelected].accessoryType = UITableViewCellAccessoryNone;
    [tableView cellForRowAtIndexPath:indexPath].accessoryType = UITableViewCellAccessoryCheckmark;

    if (indexPath.row >= [self.availabilityModel.availabilityOptions count]) {
        return;
    }
    self.lastSelected = indexPath;

    [SVProgressHUD showWithStatus:NSLocalizedString(@"Saving availability...", nil)];
    [self.availabilityModel saveUserDestination:self.lastSelected.row withCompletion:^(NSString *localizedErrorString) {
        [SVProgressHUD dismiss];
        if (localizedErrorString) {
            UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Error", nil)
                                                                           message:localizedErrorString
                                                              andDefaultButtonText:NSLocalizedString(@"Ok", nil)];
            [self presentViewController:alert
                               animated:YES
                             completion:nil];
        } else {
            [self.delegate availabilityViewController:self availabilityHasChanged:self.availabilityModel.availabilityOptions];
            [self.parentViewController dismissViewControllerAnimated:YES completion:nil];
        }
    }];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:AvailabilityAddFixedDestinationSegue]) {
        if ([segue.destinationViewController isKindOfClass:[VialerWebViewController class]]) {
            VialerWebViewController *webController = segue.destinationViewController;

            [VialerGAITracker trackScreenForControllerWithName:[VialerGAITracker GAAddFixedDestinationWebViewTrackingName]];
            webController.title = NSLocalizedString(@"Add destination", nil);
            NSString *nextURL = [NSString stringWithFormat:AvailabilityViewControllerAddFixedDestinationPageURLWithVariableForClient,
                                 [SystemUser currentUser].clientID];
            [webController nextUrl:nextURL];

        } else {
            VialerLogWarning(@"Could not segue, destinationViewController is not a \"VialerWebViewController\"");
        }
    }
}

- (IBAction)cancelButtonPressed:(UIBarButtonItem *)sender {
    [self.parentViewController dismissViewControllerAnimated:YES completion:nil];
}
@end

File: Vialer/Availability/Models/AvailabilityViewController.swift
//
//  AvailabilityViewController.swift
//  Vialer
//
//  Created by Chris Kontos on 09/10/2018.
//  Copyright © 2018 VoIPGRID. All rights reserved.
//

@objc protocol AvailabilityViewControllerDelegate: NSObjectProtocol {
    func availabilityViewController(_ controller: AvailabilityViewController?, availabilityHasChanged availabilityOptions: NSArray)
}

private let AvailabilityAddFixedDestinationSegue = "AddFixedDestinationSegue"
private let AvailabilityViewControllerAddFixedDestinationPageURLWithVariableForClient = "/client/%@/fixeddestination/add/"

@objc class AvailabilityViewController: UITableViewController{

    private var lastSelected: IndexPath?
    lazy private var availabilityModel: AvailabilityModel = {
        return AvailabilityModel()
    }()
    @objc weak var delegate: AvailabilityViewControllerDelegate?

    override func awakeFromNib() {
        super.awakeFromNib()
        refreshControl?.attributedTitle = NSAttributedString(string: NSLocalizedString("Loading availability options...", comment: ""))
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        VialerGAITracker.trackScreenForController(name: NSStringFromClass(type(of: self).self))
        
        refreshControl?.beginRefreshing()
        tableView.setContentOffset(CGPoint(x: 0, y: -(refreshControl?.frame.size.height ?? 0.0)), animated: true)
        if let unwrappedRefreshControl = refreshControl {
            loadUserDestinations(unwrappedRefreshControl)
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        tableView.tintColor = Configuration.default().colorConfiguration.color(forKey: ConfigurationAvailabilityTableViewTintColor)
    }
    
    @IBAction func loadUserDestinations(_ sender: UIRefreshControl) {
        availabilityModel.getUserDestinations({ localizedErrorString in
            if localizedErrorString != nil {
                self.present(UIAlertController(title: NSLocalizedString("Error", comment: ""), message: localizedErrorString, andDefaultButtonText: NSLocalizedString("Ok", comment: "")), animated: true)
            } else {
                self.tableView.reloadData()
            }
            self.refreshControl?.endRefreshing()
        })
    }
    
    // MARK: - Table view data source
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return availabilityModel.availabilityOptions.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let DefaultCellIdentifier = "AvailabilityTableViewDefaultCell"
        guard let cell = self.tableView.dequeueReusableCell(withIdentifier: DefaultCellIdentifier) else {
            let cell = UITableViewCell(style: .default, reuseIdentifier: DefaultCellIdentifier)
            return cell
        }
        if let availabilityDict = availabilityModel.availabilityOptions[indexPath.row] as? NSDictionary {
            if (availabilityDict[AvailabilityModelPhoneNumberKey] as? Int == 0) {
                cell.textLabel?.text = availabilityDict[AvailabilityModelDescription] as? String
            } else {
                var phoneNumber = "\(availabilityDict[AvailabilityModelPhoneNumberKey] ?? "")"
                if phoneNumber.count > 5 {
                    phoneNumber = "+" + (phoneNumber)
                }
                cell.textLabel?.text = "\(phoneNumber) / \(availabilityDict[AvailabilityModelDescription] ?? "")"
            }
            if availabilityDict[AvailabilityModelSelected] as! Int == 1 {
                cell.accessoryType = .checkmark
                lastSelected = indexPath
            } else {
                cell.accessoryType = .none
            }
        }
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        if let lastSelectedUnwrapped = lastSelected {
            tableView.cellForRow(at: lastSelectedUnwrapped)?.accessoryType = .none
        }
        tableView.cellForRow(at: indexPath)?.accessoryType = .checkmark

        if indexPath.row >= availabilityModel.availabilityOptions.count {
            return
        }
        lastSelected = indexPath
        
        SVProgressHUD.show(withStatus: NSLocalizedString("Saving availability...", comment: ""))
        availabilityModel.saveUserDestination(indexPath.row, withCompletion: { localizedErrorString in
            SVProgressHUD.dismiss()
            if localizedErrorString != nil {
                if let alert = UIAlertController(title: NSLocalizedString("Error", comment: ""), message: localizedErrorString, andDefaultButtonText: NSLocalizedString("Ok", comment: "")) {
                    self.present(alert, animated: true)
                }
            } else {
                self.delegate?.availabilityViewController(self, availabilityHasChanged: self.availabilityModel.availabilityOptions)
                self.parent?.dismiss(animated: true)
            }
        })
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if (segue.identifier == AvailabilityAddFixedDestinationSegue) {
            if (segue.destination is VialerWebViewController) {
                let webController = segue.destination as? VialerWebViewController
                
                VialerGAITracker.trackScreenForController(name: VialerGAITracker.GAAddFixedDestinationWebViewTrackingName())
                webController?.title = NSLocalizedString("Add destination", comment: "")
                let nextURL = String(format: AvailabilityViewControllerAddFixedDestinationPageURLWithVariableForClient, SystemUser.current().clientID)
                webController?.nextUrl(nextURL)
            } else {
                VialerLogWarning("Could not segue, destinationViewController is not a \"VialerWebViewController\"")
            }
        }
    }
    
    @IBAction func cancelButtonPressed(_ sender: UIBarButtonItem) {
        parent?.dismiss(animated: true)
    }
}
